# Theoritical Questions|

### 1.  What is the difference between a function and a method in Python?

### Function: 

A function is a block of code that is called by its name. It can accept parameters and may return a value. Functions in Python are defined using the 'def' keyword and are generally independent and not tied to any specific object or class.

### Method: 

A method is a function that belongs to an object (instance of a class) or a class itself. Methods are defined within a class using the 'def' keyword and are accessed using dot notation (.) on an object instance or class. Methods have access to the data (attributes) of the object they belong to through the self parameter for instance methods.

In [22]:
## Example of function
def add(a,b):
    return a+b
result= add(10,20)
print(result)



30


In [23]:

## Example of a method
class Summation:

    def add(self, a, b):
        return a + b


obj = Summation()
result = obj.add(20, 40)
print(result)

60


# 2. Explain the concept of function arguments and parameters in Python.

### Parameters:

Parameters are variables used as placeholder listed in the function definition(1st line of function).
They specify what kind of arguments a function can accept when it is called. Parameters are like variables  are used to receive input values from the caller.

In [1]:
def greetings(Name):
    print(f" Hello, {name}!, welcome to office!!!")

# Here, 'Name' is the parameter of the 'greetings' function


### Arguments:

Arguments are the actual values or variables that are passed to the function when it is called. They represent the data (or variables) we send into the function's parameters.

In [4]:
def product(a, b):
    return a * b

# Here, 'a' and 'b' are parameters of the 'product' function

result = product(10,10)

# Here, 10 and 10 are arguments passed to the 'product' function
print(result)

100


# 3. What are the different ways to define and call a function in Python?

In [15]:
"""type 1:
Basic Function Definition:"""
def greetings(Name):
    print(f"Hello, {Name}!")

greetings("Rohan")


Hello, Rohan!


In [9]:
"""type 2:
Function with Default Arguments:
"""
def product(a, b=10):
    print(a*b)

product(50)
product(50,20)


500
1000


In [13]:
"""type 3
Function with Arbitrary Arguments:
"""
def add(*args):
    s = 0
    for num in args:
        s += num
    return s

print(add(1, 2, 2,3, 4,5,6,7,8,9,10,11,12,20))


100


In [16]:
"""type 3
Function with Keyword Arguments:
"""
def print_info(**kwargs):
    for key, value in kwargs.items():
        print(f"{key}: {value}")

print_info(name="Arnab Lala", age=50, city="Siliguri")


name: Arnab Lala
age: 50
city: Siliguri


In [17]:
"""type 3
Anonymous Functions: we implement anonymous func using 'lambda keyword'
"""
add = lambda a,b: a+b
print(add(6,6))


12


# 4. What is the purpose of the `return` statement in a Python function?

### Return

is a jump statement in python programming language.
If inside a function, return is invoked then the function returns the value calculated, and terminates the program.
It can also be used to arbitarily terminate the program.


In [19]:
def add_numbers(a, b):
    return a + b

result = add_numbers(100,100)
print(result)  


200


In [21]:
def is_even(number):
    if number % 2 == 0:
        return True
    else:
        print("Showing program termination")
        return False

print(is_even(5))  
print(is_even(100))  


Showing program termination
False
True


In [24]:
"retuning  multiple values as tuple "
def square_and_cube(number):
    return number ** 2, number ** 3

square, cube = square_and_cube(3)
print(square) 
print(cube)    


9
27


# 5. What are iterators in Python and how do they differ from iterables?

iterators:
anything that cn be iterated is called iterator

itererable : it is any python object/ sequenced data structure i.e capable of returning value sone by one. permitting it to iterate over a loop. eg- tuple/ list/ dict.

In [56]:
# example
lst=iter([1,2,3,4,5])
iter(lst)
print(next(lst))
print(next(lst))
print(next(lst))
print(next(lst))
print(next(lst))


1
2
3
4
5


# 6. Explain the concept of generators in Python and how they are defined.

Generators:  Generators are a simple way to create iterators using functions and the yield keyword instead of returning values.

In [74]:
def countdown(num):
    n=num
    while n>0:
        yield n
        n-=1

## creating generetor 
gen =countdown(10)

##printing its values
for g in gen:
    print(g)

10
9
8
7
6
5
4
3
2
1


# 7. What are the advantages of using generators over regular functions?

I here note down various advantages of using generators over regular function:

1.On-Demand Data: Generators produce values one at a time and only when required, avoiding the need to generate or store all values upfront.

2.Memory Efficiency: Since values are generated as needed and not all at once, generators use minimal memory.

3.Efficient Memory Usage: Generators do not store the entire sequence in memory, making them suitable for large datasets or infinite sequences.


4.Yield vs. Return: Unlike the return statement, which exits the function, yield returns a value and pauses the function’s state, allowing it to resume from where it left off.

5.Multiple Yields: Functions can yield multiple values over time, producing a sequence of results.

In [89]:
## creating regular function

def regular_countDown(num):
    cd=[]
    n=num
    while n>0:
        cd.append(n) ## Here the values are actually stored and maintained in a list.. consuming enough memory
        n-=1
    return cd

result=regular_countDown(5)
for i in result:
    print(i)
print("-------------------")

## creating generator function

def generator_countdownn(num):
    n=num
    while n>0:
        yield n ##Here the values ain't stored like in list. but invoked only when called.
        n-=1

## creating generetor 
gen =countdown(5)

##printing its values
for g in gen:
    print(g)

5
4
3
2
1
-------------------
5
4
3
2
1


# 8. What is a lambda function in Python and when is it typically used?

Lambda functions are small, anonymous functions defined using the lambda keyword. They are used for creating small, throwaway functions without the need to formally define a function using def

they can contain only one expression, which is evaluated and returned.

Mainly used the code complexity is less and line of code can be shortend

In [88]:
subtract = lambda x,y:x-y 
print(subtract(50,10))

40


# 9. Explain the purpose and usage of the `map()` function in Python.

The map function applies a given function to all items in an input iterable (like a list) and returns an iterator with the results.

Used to transform each item in an iterable by applying the specified function.

In [90]:
## code in map()
lst=[1,2,3,4,5]
squared = map(lambda x:x**x, lst)

for s in squared:
    print(s)

1
4
27
256
3125


# 10. What is the difference between `map()`, `reduce()`, and `filter()` functions in Python?

##### map ():
The map function applies a given function to all items in an input iterable (like a list) and returns an iterator with the results.

Used to transform each item in an iterable by applying the specified function.

syntax:
map(<function name> ,iterable)

In [91]:
## code in map()
lst=[1,2,3,4,5]
squared = map(lambda x:x**x, lst)

for s in squared:
    print(s)

1
4
27
256
3125


#### reduce():
The reduce function from the functools module applies a given function cumulatively to the items of a sequence, from left to right, to reduce the sequence to a single value.

 It reduces an iterable to a single cumulative value by applying the function cumulatively.
 
syntax : 

from functools import reduce

reduce(function, iterable)

In [95]:
## code in reduce()
from functools import reduce
lst=[1,2,3,4,5]
add = reduce(lambda x,y:x+y, lst)

print(add)

15


#### Filter Functions()
The filter function constructs an iterator from elements of an iterable for which a specified function returns True.
 
Used for filtering elements from an iterable based on a condition defined by the function. 

syntax:
filter(function, iterable)

In [102]:
## code in filter()
num=[-5,-4,4,5]
positive= filter(lambda x:x>0, num)
print(tuple(positive))

(4, 5)


## 11. Using pen & Paper write the internal mechanism for sum operation using  reduce function on this given
list:[47,11,42,13]?

In [2]:
lst=[47,11,42,13]

from functools import reduce
result = reduce(lambda x,y:x+y, lst)
print(result)

113


### here, from functools import reduce: this statement imports the reduce function from the functools module.

lst = [47, 11, 42, 13]:  the list of numbers to calculate sum

result = reduce(lambda x,y:x+y, lst) ## this statement implememts the reduce function . i have take lambda func to calculate sum in one go. it also takes the iterable data structure i.e list. reduce()==> applies add cumulatively to the items of lst, from left to right, so it effectively computes ((47 + 11) + 42) + 13.

print(result) ## store sthe added value in variable result and print it.


# Practical Questions

### 1. Write a Python function that takes a list of numbers as input and returns the sum of all even numbers in the list.

In [1]:

lst=[]
lngth=int(input("Enter the size of list::\n"))
for i in range(lngth):
    num= int(input("Enter the Numbers in the list::\n"))
    lst.append((num))


def add(l):
    s=0
    for i in l:
        if i % 2 == 0:
            s += i
    return (s)
addition=add(lst)
print("The Sum of all even numbers in list are::",addition)

Enter the size of list::
 5
Enter the Numbers in the list::
 2
Enter the Numbers in the list::
 -6
Enter the Numbers in the list::
 56
Enter the Numbers in the list::
 13
Enter the Numbers in the list::
 -13


The Sum of all even numbers in list are:: 52


### 2. Create a Python function that accepts a string and returns the reverse of that string.

In [2]:
input_string= input("Enter a String to reverse::")

def rev_String(input_string):
    input_string=input_string[::-1]

    return input_string
reversed_string=rev_String(input_string)
print(reversed_string)

Enter a String to reverse:: Hello, Programmers!


!sremmargorP ,olleH


### 3. Implement a Python function that takes a list of integers and returns a new list containing the squares of each number.

In [6]:
from math import pow

lst=[]
lngth=int(input("Enter the size of list::\n"))
for i in range(1,lngth+1):
    num= int(input("Enter the Numbers in the list::\n"))
    lst.append((num))

def sqrd_List(lst):
    new_lst=[]
    for i in lst:
        new_lst.append(pow(i,2))
    return new_lst

result=sqrd_List(lst)
print(result)

Enter the size of list::
 5
Enter the Numbers in the list::
 1
Enter the Numbers in the list::
 2
Enter the Numbers in the list::
 3
Enter the Numbers in the list::
 4
Enter the Numbers in the list::
 5


[1.0, 4.0, 9.0, 16.0, 25.0]


### 4. Write a Python function that checks if a given number is prime or not from 1 to 200.

In [ ]:
## prgramme to check if a number is prime or not in range 1 to 200

limit = int(input("Enter the range::\n"))
def check_prime(limit):

    i=1
    while i<=limit:
        ct=0
        j=1
        while j<=i:
            ##print(i,j)
            if i%j == 0:
                ct+=1
            j+=1
        if(ct==2):
            print(i ,"is a prime nos!!!")
        else:
            print(i, "is not a prime nos!!!")
        i+=1


check_prime(limit)

## output
E:\PythonBaseFiles\python.exe "E:\PW 5th module practical solutions\Question_1.py" 

Enter the range::

200

1 is not a prime nos!!!

2 is a prime nos!!!

3 is a prime nos!!!

4 is not a prime nos!!!

5 is a prime nos!!!

6 is not a prime nos!!!

7 is a prime nos!!!

8 is not a prime nos!!!

9 is not a prime nos!!!

10 is not a prime nos!!!

11 is a prime nos!!!

12 is not a prime nos!!!

13 is a prime nos!!!

14 is not a prime nos!!!


15 is not a prime nos!!!

16 is not a prime nos!!!

17 is a prime nos!!!

18 is not a prime nos!!!

19 is a prime nos!!!

20 is not a prime nos!!!

21 is not a prime nos!!!

22 is not a prime nos!!!

23 is a prime nos!!!

24 is not a prime nos!!!

25 is not a prime nos!!!

26 is not a prime nos!!!

27 is not a prime nos!!!

28 is not a prime nos!!!

29 is a prime nos!!!

30 is not a prime nos!!!

31 is a prime nos!!!

32 is not a prime nos!!!

33 is not a prime nos!!!

34 is not a prime nos!!!

35 is not a prime nos!!!

36 is not a prime nos!!!

37 is a prime nos!!!

38 is not a prime nos!!!

39 is not a prime nos!!!

40 is not a prime nos!!!

41 is a prime nos!!!

42 is not a prime nos!!!

43 is a prime nos!!!

44 is not a prime nos!!!

45 is not a prime nos!!!

46 is not a prime nos!!!

47 is a prime nos!!!

48 is not a prime nos!!!

49 is not a prime nos!!!

50 is not a prime nos!!!

51 is not a prime nos!!!

52 is not a prime nos!!!

53 is a prime nos!!!

54 is not a prime nos!!!

55 is not a prime nos!!!

56 is not a prime nos!!!

57 is not a prime nos!!!

58 is not a prime nos!!!

59 is a prime nos!!!

60 is not a prime nos!!!

61 is a prime nos!!!

62 is not a prime nos!!!

63 is not a prime nos!!!

64 is not a prime nos!!!

65 is not a prime nos!!!

66 is not a prime nos!!!

67 is a prime nos!!!

68 is not a prime nos!!!


69 is not a prime nos!!!

70 is not a prime nos!!!

71 is a prime nos!!!

72 is not a prime nos!!!


73 is a prime nos!!!


74 is not a prime nos!!!

75 is not a prime nos!!!

76 is not a prime nos!!!

77 is not a prime nos!!!

78 is not a prime nos!!!

79 is a prime nos!!!


80 is not a prime nos!!!

81 is not a prime nos!!!

82 is not a prime nos!!!

83 is a prime nos!!!

84 is not a prime nos!!!

85 is not a prime nos!!!

86 is not a prime nos!!!

87 is not a prime nos!!!

88 is not a prime nos!!!

89 is a prime nos!!!

90 is not a prime nos!!!

91 is not a prime nos!!!

92 is not a prime nos!!!

93 is not a prime nos!!!

94 is not a prime nos!!!

95 is not a prime nos!!!

96 is not a prime nos!!!

97 is a prime nos!!!

98 is not a prime nos!!!

99 is not a prime nos!!!

100 is not a prime nos!!!

101 is a prime nos!!!

102 is not a prime nos!!!

103 is a prime nos!!!

104 is not a prime nos!!!

105 is not a prime nos!!!

106 is not a prime nos!!!

107 is a prime nos!!!

108 is not a prime nos!!!

109 is a prime nos!!!

110 is not a prime nos!!!

111 is not a prime nos!!!

112 is not a prime nos!!!

113 is a prime nos!!!

114 is not a prime nos!!!

115 is not a prime nos!!!

116 is not a prime nos!!!

117 is not a prime nos!!!

118 is not a prime nos!!!

119 is not a prime nos!!!

120 is not a prime nos!!!

121 is not a prime nos!!!

122 is not a prime nos!!!

123 is not a prime nos!!!

124 is not a prime nos!!!

125 is not a prime nos!!!

126 is not a prime nos!!!

127 is a prime nos!!!

128 is not a prime nos!!!

129 is not a prime nos!!!

130 is not a prime nos!!!

131 is a prime nos!!!

132 is not a prime nos!!!

133 is not a prime nos!!!

134 is not a prime nos!!!

135 is not a prime nos!!!

136 is not a prime nos!!!

137 is a prime nos!!!

138 is not a prime nos!!!

139 is a prime nos!!!

140 is not a prime nos!!!

141 is not a prime nos!!!

142 is not a prime nos!!!

143 is not a prime nos!!!

144 is not a prime nos!!!

145 is not a prime nos!!!

146 is not a prime nos!!!

147 is not a prime nos!!!

148 is not a prime nos!!!

149 is a prime nos!!!

150 is not a prime nos!!!

151 is a prime nos!!!

152 is not a prime nos!!!

153 is not a prime nos!!!

154 is not a prime nos!!!

155 is not a prime nos!!!

156 is not a prime nos!!!

157 is a prime nos!!!

158 is not a prime nos!!!

159 is not a prime nos!!!

160 is not a prime nos!!!

161 is not a prime nos!!!

162 is not a prime nos!!!

163 is a prime nos!!!

164 is not a prime nos!!!

165 is not a prime nos!!!

166 is not a prime nos!!!

167 is a prime nos!!!

168 is not a prime nos!!!

169 is not a prime nos!!!

170 is not a prime nos!!!

171 is not a prime nos!!!

172 is not a prime nos!!!

173 is a prime nos!!!

174 is not a prime nos!!!

175 is not a prime nos!!!

176 is not a prime nos!!!

177 is not a prime nos!!!

178 is not a prime nos!!!

179 is a prime nos!!!

180 is not a prime nos!!!

181 is a prime nos!!!

182 is not a prime nos!!!

183 is not a prime nos!!!

184 is not a prime nos!!!

185 is not a prime nos!!!

186 is not a prime nos!!!

187 is not a prime nos!!!

188 is not a prime nos!!!

189 is not a prime nos!!!

190 is not a prime nos!!!

191 is a prime nos!!!

192 is not a prime nos!!!

193 is a prime nos!!!

194 is not a prime nos!!!

195 is not a prime nos!!!

196 is not a prime nos!!!

197 is a prime nos!!!

198 is not a prime nos!!!

199 is a prime nos!!!

200 is not a prime nos!!!

Process finished with exit code 0


## 5. Create an iterator class in Python that generates the Fibonacci sequence up to a specified number of terms

In [2]:
# code
class FibonacciIterator:
    def __init__(self, max_terms):
        self.max_terms = max_terms
        self.term_count = 0
        self.a, self.b = 0, 1
    
    def __iter__(self):
        return self
    
    def __next__(self):
        if self.term_count >= self.max_terms:
            raise StopIteration
        
        if self.term_count == 0:
            self.term_count += 1
            return 0
        
        if self.term_count == 1:
            self.term_count += 1
            return 1
        
        fibonacci_number = self.a + self.b
        self.a, self.b = self.b, fibonacci_number
        self.term_count += 1
        return fibonacci_number

# Creating an object (instance) of FibonacciIterator
fib_iter = FibonacciIterator(10)  # Generate Fibonacci sequence up to 10 terms

# Using the iterator to print Fibonacci numbers
for num in fib_iter:
    print(num)


0
1
1
2
3
5
8
13
21
34


### 6. Write a generator function in Python that yields the powers of 2 up to a given exponent.

In [1]:
max_expo=int(input("Enter the maximum exponent::\n"))
def powerOfTwo(max_expo):
    expo=0
    while(expo<=max_expo):
        yield 2**expo
        expo+=1

gen= powerOfTwo(max_expo)
print("The generative values are ::\n")
for g in gen:
    print(g)

Enter the maximum exponent::
 6


The generative values are ::

1
2
4
8
16
32
64


## 7. Implement a generator function that reads a file line by line and yields each line as a string

In [4]:
## code
def reading_file():
    f = open("sample.txt", "r")
    data=f.read()
    yield data

gen = reading_file()
print("Read each line from file and yeilding each line as a sting\n\n")
for g in gen:
    print(g)


Read each line from file and yeilding each line as a sting


Hi everyone
 We are learning File I/O
 using  Java.
 i like programming in Java



## 8. Use a lambda function in Python to sort a list of tuples based on the second element of each tuple.

In [1]:
## code
# Sample list of tuples
tuples_list = [(1, 5), (3, 2), (9, 8), (6, 4), (5, 7)]

# Sort the list of tuples based on the second element of each tuple
sorted_list = sorted(tuples_list, key=lambda x: x[1])

print(sorted_list)


[(3, 2), (6, 4), (1, 5), (5, 7), (9, 8)]


### 9.Write a Python program that uses `map()` to convert a list of temperatures from Celsius to Fahrenheit

In [6]:
lst=[5, 8.9,-9.0,100,-100,20,37.0]
# formula ==>  F=9/5 * C+32

celcius_to_fahrenheit= map(lambda celcius: 9/5 *celcius +32,lst )

print("The mapped values of CELCIUS scale with FAHARENHEIT scale  is as follows::\n") 
for f in celcius_to_fahrenheit:
    print(f)

The mapped values of CELCIUS scale with FAHARENHEIT scale  is as follows::

41.0
48.019999999999996
15.8
212.0
-148.0
68.0
98.60000000000001


### 10. Create a Python program that uses `filter()` to remove all the vowels from a given string.

In [9]:
input_string = input("Enter a String::\n")


def removed_all_vowels(input_str):
    vowels = "aeiouAEIOU"
    # Use filter to keep only characters not in vowels
    filtered_str = filter(lambda x: x not in vowels, input_str)
    # Join the filtered characters to form the resulti|ng string
    result_str = ''.join(filtered_str)
    return result_str



result_string = removed_all_vowels(input_string)
print("Original string:", input_string)
print("String without vowels:", result_string)

Enter a String::
 A quick brown fox jumped over a lazy Dog


Original string: A quick brown fox jumped over a lazy Dog
String without vowels:  qck brwn fx jmpd vr  lzy Dg
